---
title: "05_ES22_Genes_DEG"
output: html_document
date: "2024-05-28"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r library}
library("tidyverse")
library("DESeq2")
library("pheatmap")
library("vsn")
library("paletteer")
library("apeglm")
library("ggrepel")
```

```{r dataload}
setwd("~/sciebo/Dry_lab/ExStream22_IMP/transcriptomics/Gene_cluster_09/")
dat <- read_tsv("ES_samples_vs_ES22_IMP_prokarya_genes_mmseq2DB_clustered_linclust_representative_09_read_counts_coverm_no_missmatch_filtering.tsv_cut", progress = T,num_threads = 10) ## already loaded
b6 <-  read_tsv("headers.tmp_cut_min1000.genes.faa-vs-FunTaxDB1.3.b6_cut", progress = T,num_threads = 10, col_names = F) 
meta <- read.table("All_samples_meta.tsv", row.names=1, sep = "\t", header = T)


dat_stress <- dat  %>% 
  gather("sample", "coverage", -Contig) %>% 
  left_join(meta_no_index %>% rownames_to_column("sample")) %>% 
  filter(phase == "stress")%>% 
  select(sample,coverage,Contig) %>% 
  spread("sample", "coverage") %>% 
  column_to_rownames("Contig")

meta_stress <- meta %>% 
filter(phase == "stress")  %>%
  mutate(temperature = str_replace(temperature, "normal", "background"))

selected_columns_stress <- as.character(rownames(meta_stress))
dat_stress <- dat %>% 
 column_to_rownames("Contig") 
dat_stress <- dat_stress[, selected_columns_stress]
#write.table(dat_stress, file = "dat_stress.tsv", sep = "\t", row.names = T,quote = F)

dat_recovery <- dat %>% 
  rownames_to_column("Contig") %>% 
  gather("sample", "coverage", -Contig) %>% 
  left_join(meta_no_index %>% rownames_to_column("sample")) %>% 
  filter(phase == "recovery")%>% 
  select(sample,coverage,Contig) %>% 
  spread("sample", "coverage") %>% 
  column_to_rownames("Contig")

meta_recovery <- meta %>% 
filter(phase == "recovery") %>%
  mutate(temperature = str_replace(temperature, "normal", "background"))

selected_columns_recovery<- as.character(rownames(meta_recovery))
dat_recovery <- dat %>% 
 column_to_rownames("Contig") 
dat_recovery <- dat_recovery[, selected_columns_recovery]
#write.table(dat_recovery, file = "dat_recovery.tsv", sep = "\t", row.names = T,quote = F)

dat_modify <- dat %>% 
  column_to_rownames("Contig")

meta_modify <- meta %>%
    mutate(temperature = str_replace(temperature, "normal", "background"))


#dat_stress <- read.table("dat_stress.tsv", row.names=1, sep = "\t", header = T)
#dat_recovery <- read.table("dat_recovery.tsv", row.names=1, sep = "\t", header = T)

```

```{r dds_stress}
smallestGroupSize <- 3


dds_stress <- DESeqDataSetFromMatrix(countData = dat_stress,
                              colData = meta_stress,
                              design= ~  velocity * salinity * temperature)

keep_stress <- rowSums(counts(dds_stress) >= 5) >= smallestGroupSize
dds_stress <- dds_stress[keep_stress,]

dds_stress <- DESeq(dds_stress)
res_stress_velocity <- results(dds_stress, contrast=c("velocity","reduced","normal"),alpha = 0.05)
resLFC_velocity_stress <- lfcShrink(dds_stress, coef="velocity_reduced_vs_normal", type="apeglm",res = res_stress_velocity) 
resLFC_velocity_stress%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)

res_stress_salinity <- results(dds_stress, contrast=c("salinity","salt","background"),alpha = 0.05)
resLFC_salinity_stress <- lfcShrink(dds_stress, coef="salinity_salt_vs_background", type="apeglm",res = res_stress_salinity) 
resLFC_salinity_stress%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)

res_stress_temp <- results(dds_stress, contrast=c("temperature","increased","background"),alpha = 0.05)
resLFC_temp_stress <- lfcShrink(dds_stress, coef="temperature_increased_vs_background",res = res_stress_temp, type="apeglm") 
resLFC_temp_stress%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)
res_stress_velocity_salinity <- results(dds_stress, name=c("velocityreduced.salinitysalt"),alpha = 0.05)
resLFC_velocity_salinity_stress <- lfcShrink(dds_stress, coef="velocityreduced.salinitysalt", type="apeglm",res = res_stress_velocity_salinity)
resLFC_velocity_salinity_stress%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)


res_stress_velocity_temperature <- results(dds_stress, name=c("velocityreduced.temperatureincreased"),alpha = 0.05)
resLFC_velocity_temperature_stress<-lfcShrink(dds_stress, coef="velocityreduced.temperatureincreased", type="apeglm",res = res_stress_velocity_temperature) 
resLFC_velocity_temperature_stress%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)


res_stress_salt_temperature <- results(dds_stress, name=c("salinitysalt.temperatureincreased"),alpha = 0.05)
resLFC_salinity_temperature_stress <- lfcShrink(dds_stress, coef="salinitysalt.temperatureincreased", type="apeglm",res = res_stress_salt_temperature)
resLFC_salinity_temperature_stress %>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)

res_stress_all_stressors <- results(dds_stress, name=c("velocityreduced.salinitysalt.temperatureincreased"),alpha = 0.05)
resLFC_all_stressors_stress <- lfcShrink(dds_stress, coef="velocityreduced.salinitysalt.temperatureincreased", type="apeglm",res = res_stress_all_stressors)
resLFC_all_stressors_stress%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)
```

```{r dds_recovery}
smallestGroupSize <- 3


dds_recovery <- DESeqDataSetFromMatrix(countData = dat_recovery,
                              colData = meta_recovery,
                              design= ~  velocity * salinity * temperature)

keep_recovery <- rowSums(counts(dds_recovery) >= 5) >= smallestGroupSize
dds_recovery <- dds_recovery[keep_recovery,]


dds_recovery <- DESeq(dds_recovery)

res_recovery_velocity <- results(dds_recovery, contrast=c("velocity","reduced","normal"),alpha = 0.05)
resLFC_velocity_recovery <- lfcShrink(dds_recovery, coef="velocity_reduced_vs_normal", type="apeglm",res = res_recovery_velocity) 
resLFC_velocity_recovery%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)

res_recovery_salinity <- results(dds_recovery, contrast=c("salinity","salt","background"),alpha = 0.05)
resLFC_salinity_recovery <- lfcShrink(dds_recovery, coef="salinity_salt_vs_background", type="apeglm",res = res_recovery_salinity) 
resLFC_salinity_recovery%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)

res_recovery_temp <- results(dds_recovery, contrast=c("temperature","increased","background"),alpha = 0.05)
resLFC_temp_recovery <- lfcShrink(dds_recovery, coef="temperature_increased_vs_background",res = res_recovery_temp, type="apeglm") 
resLFC_temp_recovery%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)
res_recovery_velocity_salinity <- results(dds_recovery, name=c("velocityreduced.salinitysalt"),alpha = 0.05)
resLFC_velocity_salinity_recovery <- lfcShrink(dds_recovery, coef="velocityreduced.salinitysalt", type="apeglm",res = res_recovery_velocity_salinity)
resLFC_velocity_salinity_recovery%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)


res_recovery_velocity_temperature <- results(dds_recovery, name=c("velocityreduced.temperatureincreased"),alpha = 0.05)
resLFC_velocity_temperature_recovery<-lfcShrink(dds_recovery, coef="velocityreduced.temperatureincreased", type="apeglm",res = res_recovery_velocity_temperature) 
resLFC_velocity_temperature_recovery%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)


res_recovery_salt_temperature <- results(dds_recovery, name=c("salinitysalt.temperatureincreased"),alpha = 0.05)
resLFC_salinity_temperature_recovery <- lfcShrink(dds_recovery, coef="salinitysalt.temperatureincreased", type="apeglm",res = res_recovery_salt_temperature)
resLFC_salinity_temperature_recovery %>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)

res_recovery_all_stressors <- results(dds_recovery, name=c("velocityreduced.salinitysalt.temperatureincreased"),alpha = 0.05)
resLFC_all_stressors_recovery <- lfcShrink(dds_recovery, coef="velocityreduced.salinitysalt.temperatureincreased", type="apeglm",res = res_recovery_all_stressors) 
resLFC_all_stressors_recovery%>% as.data.frame() %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>%  mutate(scaffold = gsub('_[^_]*$', '', Contig)) %>%  filter(abs(log2FoldChange) > 1 & padj < 0.05)
```


```{r ggplot2_temp_stress}
alpha <- 0.05

res_taxa_added_temp_stress <- as.data.frame(resLFC_temp_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
 res_taxa_added_temp_stress$X2 <-str_extract(res_taxa_added_temp_stress$X2, ".*(?=n=)")


significant_points_temp_stress <- res_taxa_added_temp_stress %>%
  as.data.frame() %>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha)

volcano_plot_temp_stress <- ggplot(res_taxa_added_temp_stress, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_temp_stress, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_temp_stress, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "C) Temperature increase",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()

```
```{r ggplot2_velocity_stress}
alpha <- 0.05

res_taxa_added_velocity_stress <- as.data.frame(resLFC_velocity_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
res_taxa_added_velocity_stress$X2 <-str_extract(res_taxa_added_velocity_stress$X2, ".*(?=n=)")

significant_points_velocity_stress <- res_taxa_added_velocity_stress %>%
  as.data.frame() %>% 
  filter(abs(log2FoldChange) > 1  & padj < alpha)

volcano_plot_velocity_stress <- ggplot(res_taxa_added_velocity_stress, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_velocity_stress, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_velocity_stress, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "B) Velocity reduction",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()

```
```{r ggplot2_salinity_stress}
alpha <- 0.05

res_taxa_added_salinity_stress <- as.data.frame(resLFC_salinity_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
res_taxa_added_salinity_stress$X2 <-str_extract(res_taxa_added_salinity_stress$X2, ".*(?=n=)")

significant_points_salinity_stress <- res_taxa_added_salinity_stress %>%
  as.data.frame() %>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha)

volcano_plot_salinity_stress <- ggplot(res_taxa_added_salinity_stress, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_salinity_stress, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_salinity_stress, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "D) Salinity increase",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()
```

```{r stressor_combinations_stress}
res_taxa_added_velocity_salinity_stress <- as.data.frame(resLFC_velocity_salinity_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
res_taxa_added_velocity_salinity_stress$X2 <-str_extract(res_taxa_added_velocity_salinity_stress$X2, ".*(?=n=)")

res_taxa_added_velocity_temperature_stresss <- as.data.frame(resLFC_velocity_temperature_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
res_taxa_added_velocity_temperature_stresss$X2 <-str_extract(res_taxa_added_velocity_temperature_stresss$X2, ".*(?=n=)")

res_taxa_added_salinity_temperature_stress <- as.data.frame(resLFC_salinity_temperature_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
res_taxa_added_salinity_temperature_stress$X2 <-str_extract(res_taxa_added_salinity_temperature_stress$X2, ".*(?=n=)")

res_taxa_added_all_stressors_stress <- as.data.frame(resLFC_all_stressors_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>%   mutate(X2 = str_replace(X2, "^\\S* ", ""))
res_taxa_added_all_stressors_stress$X2 <-str_extract(res_taxa_added_all_stressors_stress$X2, ".*(?=n=)")



volcano_plot_velocity_salinity_stress <- ggplot(res_taxa_added_velocity_salinity_stress, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_velocity_salinity_stress, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_velocity_salinity_stress, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "E) Velocity reduction and salinity increase",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()

volcano_plot_velocity_temperature_stresss <- ggplot(res_taxa_added_velocity_temperature_stresss, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_velocity_temperature_stresss, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_velocity_temperature_stresss, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "Velocity reduction and temperature increase",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()

volcano_plot_salinity_temperature_stress <- ggplot(res_taxa_added_salinity_temperature_stress, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_salinity_temperature_stress, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_salinity_temperature_stress, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "Volcano plot",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()

volcano_plot_all_stressors_stress <- ggplot(res_taxa_added_all_stressors_stress, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(data = filter(res_taxa_added_all_stressors_stress, abs(log2FoldChange) > 1 & padj < alpha),
                         size = 3, shape = 20) +
  geom_point(data = filter(res_taxa_added_all_stressors_stress, !(abs(log2FoldChange) > 1 & padj < alpha)),
             color = "gray", # Color for non-significant points
             size = 3, shape = 20) +
  geom_hline(yintercept = -log10(alpha), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "brown") +
  geom_vline(xintercept = 0, color = "black") +
  labs(title = "F) All stressors",
       x = "Effect size: log2(fold-change)",
       y = "-log10(adjusted p-value)") +
  theme_minimal()
```

```{r showplot}
library(patchwork)

volcano_plot_salinity_stress
volcano_plot_temp_stress
volcano_plot_velocity_stress
volcano_plot_velocity_salinity_stress
volcano_plot_velocity_temperature_stresss
volcano_plot_salinity_temperature_stress
volcano_plot_all_stressors_stress

volcano_plot_velocity_stress + volcano_plot_temp_stress + volcano_plot_salinity_stress + volcano_plot_velocity_salinity_stress + volcano_plot_all_stressors_stress +
    plot_layout(ncol = 3)
ggsave("ES22_IMP_genes_09_DE_stressorphase.jpeg", width = 15, height = 10) ## For Figure

```

```{r print_table}
 stressor_phase_table <- as.data.frame(resLFC_temp_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha) %>% 
  mutate(stressor = "temperature_increase_stressor_phase") %>% bind_rows(
as.data.frame(resLFC_velocity_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha) %>% 
  mutate(stressor = "velocity_reduction_stressor_phase")) %>% bind_rows(
as.data.frame(resLFC_salinity_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1)) %>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha) %>% 
  mutate(stressor = "salinity_increase_stressor_phase")) %>% bind_rows(
as.data.frame(resLFC_velocity_salinity_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha)%>% 
  mutate(stressor = "velocity_reduction_salinity_increase_stressor_phase")) %>% bind_rows(
as.data.frame(resLFC_velocity_temperature_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha)%>% 
  mutate(stressor = "velocity_reduction_temperature_increase_stressor_phase")) %>% bind_rows(
 as.data.frame(resLFC_salinity_temperature_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha)%>% 
  mutate(stressor = "salinity_increase_temperature_increase_stressor_phase")) %>% bind_rows(
as.data.frame(resLFC_all_stressors_stress) %>% rownames_to_column("Contig") %>% left_join(b6  %>% select(X1,X2) %>% dplyr::rename(Contig =X1))%>% 
  filter(abs(log2FoldChange) > 1 & padj < alpha)%>% 
  mutate(stressor = "all_stressors_stressor_phase"))%>% dplyr::rename(FunTaxDB_annotation = X2)



write.table(stressor_phase_table, file = "ES22_IMP_genesDE_stressorphase_significant_genes.tsv", sep = "\t", row.names = F,quote = F) ### For supplementary 

```


```{r triplot}
## Grouping into classes to show differential expression in a more plastic way
#stressor_phase_table_loaded <- read_tsv("ES22_IMP_genesDE_stressorphase_significant_genes.tsv", progress = T,num_threads = 10, col_names = T)  
get_annotation_group <- function(annotation) {
  if (grepl("chaperon", annotation, ignore.case = TRUE)) {
    return("Chaperones")
  } else if (grepl("ribosom", annotation, ignore.case = TRUE)) {
    return("Ribosomal proteins")
  } else if (grepl("heat|cold|HSP", annotation, ignore.case = TRUE)) {
    return("Heat/cold-shock proteins")
  } else if (grepl("photo", annotation, ignore.case = TRUE)) {
    return("Photosystem I/II")
  } else if (grepl("translocase", annotation, ignore.case = TRUE)) {
    return("Protein translocases")
  } else if (is.na(annotation) || grepl("hypothetical|uncharacterized", annotation, ignore.case = TRUE)) {
    return("Unknown function")
  } else if (grepl("polymerase", annotation, ignore.case = TRUE)) {
    return("DNA-directed RNA polymerase subunits")
  } else if (grepl("ATP synthase", annotation, ignore.case = TRUE)) {
    return("ATP synthase subunits")
  } else if (grepl("ribulose", annotation, ignore.case = TRUE) & grepl("carboxylase", annotation, ignore.case = TRUE)) {
    return("Ribulose carboxylases")
  } else if (grepl("Elongation", annotation, ignore.case = TRUE)) {
    return("Elongation factor Tu")
  } else if (grepl("Transcription", annotation, ignore.case = TRUE)) {
    return("Transcription proteins")
  } else if (grepl("cytochrome", annotation, ignore.case = TRUE)) {
    return("Cytochromes")
  } else if (grepl("receptor", annotation, ignore.case = TRUE)) {
    return("Receptors")
  } else if (grepl("fusion", annotation, ignore.case = TRUE)) {
    return("Fusion proteins")
  } else if (grepl("transport", annotation, ignore.case = TRUE)) {
    return("Transporter proteins")
  } else if (grepl("Adenylyl-sulfate", annotation, ignore.case = TRUE)) {
    return("Adenylyl-sulfate reductases")
  } else if (grepl("fusion", annotation, ignore.case = TRUE)) {
    return("Fusion proteins")
  } else if (grepl("translation", annotation, ignore.case = TRUE)) {
    return("Translation initiation factors")
  } else if (grepl("T9SS", annotation, ignore.case = TRUE)) {
    return("T9SS type A sorting proteins")
  } else if (grepl("hydroperoxide", annotation, ignore.case = TRUE)) {
    return("Alkyl hydroperoxide reductases C")
  } else if (grepl("acyl-CoA", annotation, ignore.case = TRUE)) {
    return("Acyl-CoA dehydrogenases")
  } else if (grepl("hydroperoxide", annotation, ignore.case = TRUE)) {
    return("Alkyl hydroperoxide reductases C")
  } else if (grepl("P700", annotation, ignore.case = TRUE)) {
    return("PSI P700 apoproteins A1")
  } else {
    return("Others")  # Assign "miscellaneous" for all other cases
  }
}

grouped_stressor_phase_table_loaded <- stressor_phase_table_loaded %>%  mutate(annotation_group = sapply(FunTaxDB_annotation, get_annotation_group)) %>% select(Contig,FunTaxDB_annotation,annotation_group,stressor,log2FoldChange) 

stressor_phase_table_loaded_count <-  stressor_phase_table_loaded %>%  mutate(annotation_group = sapply(FunTaxDB_annotation, get_annotation_group)) %>% select(Contig,FunTaxDB_annotation,annotation_group,stressor,log2FoldChange) %>% dplyr::count(annotation_group)


stressor_phase_table_loaded_count_merged <- grouped_stressor_phase_table_loaded %>% left_join(stressor_phase_table_loaded_count,by = "annotation_group") %>% mutate(annotation_group_filter = ifelse(n < 5, "Others", annotation_group))


stressor_phase_table_loaded_count_merged_for_print <- stressor_phase_table_loaded %>% left_join(stressor_phase_table_loaded_count_merged %>% select(Contig,annotation_group_filter))
write.table(stressor_phase_table_loaded_count_merged_for_print, file = "~/sciebo/Dry_lab/ExStream22_IMP/transcriptomics/Gene_cluster_09/ES22_IMP_genesDE_stressorphase_significant_genes_annotation_groups.tsv", sep = "\t", row.names = F,quote = F) ### For Supplementary information


y_levels <- c(levels(factor(stressor_phase_table_loaded$stressor)), "", "", "")

stressor_phase_table_loaded_count_merged_v2 <- rbind(stressor_phase_table_loaded_count_merged, 
                                     data.frame(stressor = c("tv" ,"ts"),
                                                annotation_group_filter = rep(NA, 2),
                                                Contig = rep(NA, 2),
                                                FunTaxDB_annotation = rep(NA, 2),
                                                 annotation_group = rep(NA, 2),
                                                 n = rep(NA, 2),
                                                log2FoldChange = rep(NA, 2)))

stressor_phase_table_loaded_count_merged_v2 %>%
  mutate(neg_log2FoldChange = ifelse(log2FoldChange < 0, log2FoldChange, NA)) %>% 
  mutate(log2FoldChange = ifelse(log2FoldChange > 0, log2FoldChange, NA))


stressor_phase_table_loaded_count_merged_v3_pos <- stressor_phase_table_loaded_count_merged_v2 %>%
    mutate(neg_log2FoldChange = ifelse(log2FoldChange < 0, log2FoldChange, NA)) %>% 
    mutate(log2FoldChange = ifelse(log2FoldChange > 0, log2FoldChange, NA)) %>% 
    group_by(stressor,annotation_group_filter)  %>%
    summarise_at(vars(log2FoldChange),funs(positive_sum = sum(.,na.rm=T)))

stressor_phase_table_loaded_count_merged_v3_neg <- stressor_phase_table_loaded_count_merged_v2 %>%
    mutate(neg_log2FoldChange = ifelse(log2FoldChange < 0, log2FoldChange, NA)) %>% 
    mutate(log2FoldChange = ifelse(log2FoldChange > 0, log2FoldChange, NA)) %>% 
    group_by(stressor,annotation_group_filter)  %>%
    summarise_at(vars(neg_log2FoldChange),funs(negative_sum = sum(.,na.rm=T)))

stressor_phase_table_loaded_count_merged_v4<- stressor_phase_table_loaded_count_merged_v3_pos %>% left_join(stressor_phase_table_loaded_count_merged_v3_neg)


custom_order <- c("all_stressors_stressor_phase", "velocity_reduction_salinity_increase_stressor_phase",
                  "ts", "tv", "salinity_increase_stressor_phase", "temperature_increase_stressor_phase",
                  "velocity_reduction_stressor_phase")

custom_order_y <- c("Chaperones", "Ribosomal proteins", "Photosystem I/II", "Ribulose carboxylases", 
                  "Heat/cold-shock proteins", "Protein translocases", "DNA-directed RNA polymerase subunits", 
                   "ATP synthase subunits", "Elongation factor Tu", 
                  "Cytochromes", "Others", "Unknown function", NA)


stressor_phase_table_loaded_count_merged_v4$annotation_group_filter <- factor(stressor_phase_table_loaded_count_merged_v4$annotation_group_filter, levels = custom_order_y)

ggplot(stressor_phase_table_loaded_count_merged_v4) +
    aes(x = factor(stressor, levels = custom_order))   +
    geom_bar(
        aes(
            fill = annotation_group_filter,
            y = positive_sum
        ), stat = "identity", levels=custom_order_y)  +
    geom_bar(
        aes(
            fill = annotation_group_filter,
            y = negative_sum
        ), stat = "identity",levels=custom_order_y)  +
    scale_fill_hue(direction = 1) +
    theme_bw() +
    scale_fill_paletteer_d("cartography::multi.pal", "Functional annotation", direction = 1, dynamic = TRUE) +
    coord_flip()+
    scale_y_continuous(breaks = c(-100, 0, 200, 400, 600),
                                  limits = c(-100, 600))
ggsave("ES22_IMP_genes_09_DE_plastic_stressor_impact.png", width = 15, height = 10) ### For Figure
```